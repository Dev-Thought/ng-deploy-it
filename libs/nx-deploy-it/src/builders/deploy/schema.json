{
  "$schema": "https://json-schema.org/draft-07/schema",
  "$id": "https://json-schema.org/draft-07/schema",
  "title": "NxDeployIt Deploy infrastructure",
  "description": "Deploy the infrastructure for the app",
  "type": "object",
  "properties": {
    "noBuild": {
      "type": "boolean",
      "default": false,
      "description": "Skip build process during deployment"
    },
    "nonInteractive": {
      "type": "boolean",
      "default": false,
      "description": "Changes will apply without user input"
    }
  },
  "required": []
}
