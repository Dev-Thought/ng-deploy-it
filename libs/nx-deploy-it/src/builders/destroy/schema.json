{
  "$schema": "https://json-schema.org/draft-07/schema",
  "$id": "https://json-schema.org/draft-07/schema",
  "title": "NxDeployIt Destroy infrastructure",
  "description": "Destroy the infrastructure for the app",
  "type": "object",
  "properties": {
    "nonInteractive": {
      "type": "boolean",
      "default": false,
      "description": "Changes will apply without user input"
    }
  },
  "required": []
}
