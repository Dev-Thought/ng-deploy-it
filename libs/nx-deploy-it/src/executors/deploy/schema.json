{
  "$schema": "http://json-schema.org/schema",
  "cli": "nx",
  "title": "NxDeployIt Deploy infrastructure",
  "description": "Deploy the infrastructure for the app",
  "type": "object",
  "properties": {
    "nonInteractive": {
      "type": "boolean",
      "default": false,
      "description": "Changes will apply without user input"
    }
  },
  "required": []
}
