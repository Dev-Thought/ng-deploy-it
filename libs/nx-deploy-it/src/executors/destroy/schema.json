{
  "$schema": "http://json-schema.org/schema",
  "title": "NxDeployIt Destroy infrastructure",
  "description": "Destroy the infrastructure for the app",
  "type": "object",
  "properties": {
    "nonInteractive": {
      "type": "boolean",
      "default": false,
      "description": "Changes will apply without user input"
    }
  },
  "required": []
}
