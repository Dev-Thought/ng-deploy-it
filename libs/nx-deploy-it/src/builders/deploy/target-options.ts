import { PROVIDER } from '../../utils/provider';
import { JsonObject } from '@angular-devkit/core';

export interface DeployTargetOptions extends JsonObject {
  main: string;
  provider: PROVIDER;
  pulumi: any;
}
