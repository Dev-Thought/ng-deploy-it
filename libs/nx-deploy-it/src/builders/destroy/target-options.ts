import { PROVIDER } from '../../utils/provider';
import { JsonObject } from '@angular-devkit/core';

export interface DestroyTargetOptions extends JsonObject {
  main: string;
  provider: PROVIDER;
  pulumi: any;
}
