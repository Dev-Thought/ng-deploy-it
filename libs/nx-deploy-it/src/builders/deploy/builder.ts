import {
  BuilderContext,
  BuilderOutput,
  createBuilder
} from '@angular-devkit/architect';
import { Observable, Subject, from, of } from 'rxjs';
import { switchMap, map, tap } from 'rxjs/operators';
import { NxDeployItDeployBuilderSchema } from './schema';
import {
  getApplicationType,
  ApplicationType
} from '../../utils/application-type';
import { resolve, dirname } from 'path';
import * as ncc from '@zeit/ncc';
import { DeployTargetOptions } from './target-options';
import { spawnSync, spawn } from 'child_process';
import { getPulumiBinaryPath } from '../../utils/workspace';
import { readWorkspaceConfigPath } from '@nrwl/workspace';
import { ensureDirSync, ensureFileSync } from 'fs-extra';
import { writeFileSync } from 'fs';

function spawnStack(
  cwd: string,
  configuration: string,
  projectName: string,
  withInit = false
) {
  const args = [
    'stack',
    '--stack',
    `${configuration}-${projectName}`,
    '--cwd',
    cwd
  ];
  if (withInit) {
    args.splice(1, 0, 'init');
  }

  return spawnSync(getPulumiBinaryPath(), args, {
    env: process.env
  });
}

function createStackIfNotExist(
  cwd: string,
  configuration: string,
  projectName: string
) {
  const result = spawnStack(cwd, configuration, projectName);
  if (result.stderr && result.stderr.toString().includes('no stack named')) {
    spawnStack(cwd, configuration, projectName, true);
  }
}

function up(
  cwd: string,
  options: NxDeployItDeployBuilderSchema,
  configuration: string,
  targetOptions: DeployTargetOptions,
  distPath: string,
  projectName: string,
  context: BuilderContext
) {
  const subject = new Subject<{ success: boolean }>();
  const args = [
    'up',
    '--cwd',
    cwd,
    '--stack',
    `${configuration}-${projectName}`
  ];
  if (options.nonInteractive) {
    args.push('--non-interactive', '--yes');
  }

  if (targetOptions.pulumi) {
    for (const key in targetOptions.pulumi) {
      const value = targetOptions.pulumi[key];
      if (value) {
        args.push('-c', `${key}=${value}`);
      }
    }
  }
  args.push('-c', `distPath=${distPath}`);
  args.push('-c', `projectName=${projectName}`);
  const up = spawn(getPulumiBinaryPath(), args, {
    env: { ...process.env, PULUMI_SKIP_UPDATE_CHECK: '1' },
    stdio: 'inherit'
  });

  up.on('close', code => {
    if (code !== 0) {
      context.logger.error(`up process exited with code ${code}`);
      subject.error({ success: false });
    } else {
      subject.next({ success: true });
    }
    subject.complete();
  });

  return subject.asObservable();
}

function getProjectConfig(context: BuilderContext) {
  const workspaceConfig = readWorkspaceConfigPath();

  return workspaceConfig.projects[context.target.project];
}

// TODO: Not sure if this is the best approach to get the outputPath
function getDistributionPath(context: BuilderContext) {
  const project = getProjectConfig(context);

  return resolve(
    context.workspaceRoot,
    project.architect.build.options.outputPath
  );
}

function buildProject(context: BuilderContext): Observable<BuilderOutput> {
  context.logger.info('Build project');

  const project = getProjectConfig(context);
  const applicationType = getApplicationType(project.architect.build);
  if (
    applicationType === ApplicationType.NESTJS ||
    applicationType === ApplicationType.EXPRESS
  ) {
    const infrastructureFolder = resolve(
      context.workspaceRoot,
      project.root,
      'infrastructure'
    );
    const processCwd = process.cwd();
    process.chdir(infrastructureFolder);
    return from(
      ncc(resolve(infrastructureFolder, 'functions/main/index.ts'), {
        cache: resolve(infrastructureFolder, 'buildcache')
      })
    ).pipe(
      map(
        (buildResult: {
          code: string;
          asset: { [index: string]: { source: string } };
        }) => {
          process.chdir(processCwd);
          ensureDirSync(resolve(infrastructureFolder, 'functions/dist/main'));
          // compiled javascript
          writeFileSync(
            resolve(infrastructureFolder, 'functions/dist/main/index.js'),
            buildResult.code
          );
          // assets
          for (const file in buildResult.asset) {
            const content = buildResult.asset[file];
            ensureFileSync(
              resolve(infrastructureFolder, `functions/dist/main/${file}`)
            );
            writeFileSync(
              resolve(infrastructureFolder, `functions/dist/main/${file}`),
              content.source.toString()
            );
          }
          return { success: true };
        }
      )
    );
  } else {
    return from(
      context.scheduleTarget({
        target: 'build',
        project: context.target.project,
        configuration: context.target.configuration || ''
      })
    ).pipe(switchMap(builderRun => from(builderRun.result)));
  }
}

export function runBuilder(
  options: NxDeployItDeployBuilderSchema,
  context: BuilderContext
): Observable<BuilderOutput> {
  const configuration = context.target.configuration || 'dev';

  return of({ success: true }).pipe(
    options.noBuild ? tap() : switchMap(() => buildProject(context)),
    switchMap(() =>
      from(context.getTargetOptions(context.target)).pipe(
        switchMap((targetOptions: DeployTargetOptions) => {
          const cwd = dirname(
            resolve(context.workspaceRoot, targetOptions.main as string)
          );

          createStackIfNotExist(cwd, configuration, context.target.project);

          const distributationPath = getDistributionPath(context);
          return up(
            cwd,
            options,
            configuration,
            targetOptions,
            distributationPath,
            context.target.project,
            context
          );
        })
      )
    )
  );
}

export default createBuilder(runBuilder);
