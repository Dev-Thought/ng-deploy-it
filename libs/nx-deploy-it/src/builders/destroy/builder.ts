import {
  BuilderContext,
  BuilderOutput,
  createBuilder
} from '@angular-devkit/architect';
import { Observable, from, of } from 'rxjs';
import { switchMap } from 'rxjs/operators';
import { NxDeployItDestroyBuilderSchema } from './schema';
import { spawnSync } from 'child_process';
import { getPulumiBinaryPath } from '../../utils/workspace';
import { DestroyTargetOptions } from './target-options';
import { dirname, resolve } from 'path';

function down(
  cwd: string,
  options: NxDeployItDestroyBuilderSchema,
  configuration: string,
  projectName: string
): Observable<BuilderOutput> {
  const args = [
    'destroy',
    '--cwd',
    cwd,
    '--stack',
    `${configuration}-${projectName}`
  ];
  if (options.nonInteractive) {
    args.push('--non-interactive', '--yes');
  }
  const up = spawnSync(getPulumiBinaryPath(), args, {
    env: { ...process.env, PULUMI_SKIP_UPDATE_CHECK: '1' },
    stdio: 'inherit'
  });

  if (up.error) {
    return of({ success: false, error: up.error.message });
  }

  return of({ success: true });
}

export function runBuilder(
  options: NxDeployItDestroyBuilderSchema,
  context: BuilderContext
): Observable<BuilderOutput> {
  if (!context.target && !context.target.target) {
    return of({ success: false });
  }

  const configuration = context.target.configuration || 'dev';

  return from(context.getTargetOptions(context.target)).pipe(
    switchMap((targetOptions: DestroyTargetOptions) => {
      const cwd = dirname(
        resolve(context.workspaceRoot, targetOptions.main as string)
      );

      return down(cwd, options, configuration, context.target.project);
    })
  );
}

export default createBuilder(runBuilder);
