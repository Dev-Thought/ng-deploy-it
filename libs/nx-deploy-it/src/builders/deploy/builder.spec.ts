import { NxDeployItDeployBuilderSchema } from './schema';
import { MockBuilderContext } from '@nrwl/workspace/testing';
import * as childProcess from 'child_process';
import { getMockContext } from '../../utils-test/builders.utils';
import { runBuilder } from './builder';
import { DeployTargetOptions } from './target-options';
import * as nrwlWorkspce from '@nrwl/workspace';
import * as utils from '../../utils/application-type';
import * as ncc from '@zeit/ncc';
import * as fsExtra from 'fs-extra';
import * as fs from 'fs';
jest.mock('@zeit/ncc');

const options: NxDeployItDeployBuilderSchema = {};

describe('Command Runner Builder - Deploy', () => {
  let context: MockBuilderContext;
  let options: NxDeployItDeployBuilderSchema;
  const spawnSync = jest.spyOn(childProcess, 'spawnSync');

  beforeEach(async () => {
    context = await getMockContext();

    options = {
      project: 'project-mock'
    };

    spawnSync.mockReturnValue({
      pid: null,
      output: [''],
      error: null,
      signal: null,
      status: null,
      stderr: null,
      stdout: null
    });

    await context.addTarget(
      { project: 'project-mock', configuration: 'dev', target: 'deploy' },
      'deploy'
    );
    context.target = {
      project: 'project-mock',
      configuration: 'dev',
      target: 'deploy'
    };
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should run deploy for a react app', async () => {
    jest.spyOn(context, 'getTargetOptions').mockResolvedValue({
      main: 'somewhere.ts'
    } as DeployTargetOptions);
    const scheduleTargetSpy = jest
      .spyOn(context, 'scheduleTarget')
      .mockResolvedValue({
        result: new Promise(resolve => resolve({ success: true }))
      } as any);
    jest.spyOn(nrwlWorkspce, 'readWorkspaceConfigPath').mockReturnValue({
      projects: {
        'project-mock': {
          architect: {
            build: {
              builder: '@nrwl/web:build',
              options: {
                main: 'apps/project-mock/src/main.tsx',
                webpackConfig: '@nrwl/react/plugins/webpack',
                outputPath: 'dist/apps/project-mock'
              }
            }
          }
        }
      }
    });
    const result = await runBuilder(options, context).toPromise();

    expect(scheduleTargetSpy.mock.calls[0]).toMatchSnapshot(
      'build schedule target'
    );
    expect(spawnSync.mock.calls[0][1]).toMatchSnapshot(
      'create stack if not exists'
    );
    expect(spawnSync.mock.calls[1][1]).toMatchSnapshot('deploy with pulumi');
    expect(result).toMatchSnapshot('Result of the pulumi script');
  });

  it('should update pulumi properties', async () => {
    jest.spyOn(context, 'getTargetOptions').mockResolvedValue({
      main: 'somewhere.ts',
      pulumi: {
        'aws:region': 'eu-central-1'
      }
    } as DeployTargetOptions);
    jest.spyOn(context, 'scheduleTarget').mockResolvedValue({
      result: new Promise(resolve => resolve({ success: true }))
    } as any);
    jest.spyOn(nrwlWorkspce, 'readWorkspaceConfigPath').mockReturnValue({
      projects: {
        'project-mock': {
          architect: {
            build: {
              builder: '@nrwl/web:build',
              options: {
                main: 'apps/project-mock/src/main.tsx',
                webpackConfig: '@nrwl/react/plugins/webpack',
                outputPath: 'dist/apps/project-mock'
              }
            }
          }
        }
      }
    });

    await runBuilder(options, context).toPromise();

    expect(spawnSync.mock.calls[1][1]).toMatchSnapshot('deploy with pulumi');
  });

  it('should deploy with custom build (ncc)', async () => {
    jest.spyOn(context, 'getTargetOptions').mockResolvedValue({
      main: 'somewhere.ts',
      pulumi: {
        'aws:region': 'eu-central-1'
      }
    } as DeployTargetOptions);
    jest.spyOn(context, 'scheduleTarget').mockResolvedValue({
      result: new Promise(resolve => resolve({ success: true }))
    } as any);
    jest
      .spyOn(utils, 'getApplicationType')
      .mockReturnValueOnce(utils.ApplicationType.NESTJS);
    jest.spyOn(nrwlWorkspce, 'readWorkspaceConfigPath').mockReturnValue({
      projects: {
        'project-mock': {
          architect: {
            build: {
              builder: '@nrwl/node:build',
              options: {
                outputPath: 'dist/apps/api',
                main: 'apps/api/src/main.ts',
                tsConfig: 'apps/api/tsconfig.app.json',
                assets: ['apps/api/src/assets']
              }
            }
          },
          root: 'apps/api'
        }
      }
    });
    jest.spyOn(process, 'chdir').mockReturnValue();
    jest.spyOn(process, 'cwd').mockReturnValue('mockProcessCwd');
    (ncc as jest.SpyInstance).mockResolvedValueOnce({
      code: 'Some Code',
      asset: { 'asset1.png': { source: 'code of asset 1' } }
    });
    jest.spyOn(fsExtra, 'ensureDirSync').mockReturnThis();
    jest.spyOn(fsExtra, 'ensureFileSync').mockReturnThis();
    jest.spyOn(fs, 'writeFileSync').mockReturnThis();

    await runBuilder(options, context).toPromise();

    expect(process.cwd).toHaveBeenCalled();
    expect(process.chdir).toHaveBeenCalledWith('/root/apps/api/infrastructure');
    expect(process.chdir).toHaveBeenCalledWith('mockProcessCwd');

    // check if build was written
    expect(fsExtra.ensureDirSync).toHaveBeenCalledWith(
      '/root/apps/api/infrastructure/functions/dist/main'
    );
    expect(fs.writeFileSync).toHaveBeenCalledWith(
      '/root/apps/api/infrastructure/functions/dist/main/index.js',
      'Some Code'
    );

    // check if assets were written
    expect(fsExtra.ensureFileSync).toHaveBeenCalledWith(
      '/root/apps/api/infrastructure/functions/dist/main/asset1.png'
    );
    expect(fs.writeFileSync).toHaveBeenCalledWith(
      '/root/apps/api/infrastructure/functions/dist/main/asset1.png',
      'code of asset 1'
    );
  });
});
