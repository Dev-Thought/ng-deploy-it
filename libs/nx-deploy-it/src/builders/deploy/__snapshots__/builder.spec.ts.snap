// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Command Runner Builder - Deploy should run deploy for a react app: Result of the pulumi script 1`] = `
Object {
  "success": true,
}
`;

exports[`Command Runner Builder - Deploy should run deploy for a react app: build schedule target 1`] = `
Array [
  Object {
    "configuration": "dev",
    "project": "project-mock",
    "target": "build",
  },
]
`;

exports[`Command Runner Builder - Deploy should run deploy for a react app: create stack if not exists 1`] = `
Array [
  "stack",
  "--stack",
  "dev-project-mock",
  "--cwd",
  "/root",
]
`;

exports[`Command Runner Builder - Deploy should run deploy for a react app: deploy with pulumi 1`] = `
Array [
  "up",
  "--cwd",
  "/root",
  "--stack",
  "dev-project-mock",
  "-c",
  "distPath=/root/dist/apps/project-mock",
  "-c",
  "projectName=project-mock",
]
`;

exports[`Command Runner Builder - Deploy should update pulumi properties: deploy with pulumi 1`] = `
Array [
  "up",
  "--cwd",
  "/root",
  "--stack",
  "dev-project-mock",
  "-c",
  "aws:region=eu-central-1",
  "-c",
  "distPath=/root/dist/apps/project-mock",
  "-c",
  "projectName=project-mock",
]
`;
